---
title: "Motif DB stats"
author: "Frank (zhangc518@gmail.com)"
date: "`r date()`"
always_allow_html: yes
output:
   md_document:
    variant: 
      markdown_github
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = FALSE)
require(tidyverse)

jaspar <- read.table("./db/jaspar2018_vertebrates.txt",skip = 1,stringsAsFactors = F)[,2:3] #579
tfclass <- readRDS("./db/tfclass.rds") 
dic<- readRDS("./db/dic_jaspar_tfclass.rds")
```

# jaspar stats 

The `JASPAR motifs`  were downloaded from [Jaspar](http://jaspar.genereg.net/downloads/) choosing `JASPAR2018_CORE_vertebrates_non-redundant` in `MEME` format. Because there is no gene ids provided in the Jaspar database, the corresponding gene's `Ensembl ID` was fatched by querying the provided `UniprotID` from [UniprotID](https://www.uniprot.org/uniprot/) using a customized function `getEnsemblFromJaspar` [code](./aux_funs.R). 

* number of TF motifs in Jaspar 
```{r}
nrow(jaspar) 
```
* How many of them are protein complex? 
```{r}
length(grep(":",jaspar$V3)) 
```

# tfclass stats 

[TFclass database](http://tfclass.bioinf.med.uni-goettingen.de/about.jsf) were downloaded (2018-05-14), [parsed](./scripts/parseTfclassTTL.sh) and [extacted](./scripts/tfclass.R) the human and mouse genes' Ensembl ID. 

* How many unique genus (i.e. TF genes) in parsed TFClass?
```{r}
length(unique(tfclass$merge$genus.id)) 
```
* How many subfamilies in parsed TFClass?
```{r}
length(unique(tfclass$merge$subfamily.id))
```
* How many families in parsed TFClass?
```{r}
length(unique(tfclass$merge$family.id))
```
* There are two tfs have no family info in TFClass db:
```{r}
knitr::kable(tfclass$merge %>% filter(subfamily.id==""))
```

# jaspar-tfclass dictionary stats 

`JASPAR` and `TFClass` were majorly [mapped](./cmpDB.R) by matching `Ensembl ID` except `20` TFs by `TF Name` directly. 

## how many Jaspar motif have annotation? 
```{r}
a=nrow(dic$jasparTOtfclass %>% 
  dplyr::select(jaspar.id,genus.id) %>% 
  group_by(jaspar.id)%>%
  summarise(genus.id.2 = paste0(genus.id,collapse = ';')))
paste(a,nrow(jaspar),sep = "/")
```

## which jaspar motif don't have annotation?
```{r}
noAnno <- subset(jaspar, V2 %in% setdiff(jaspar$V2,dic$jasparTOtfclass$jaspar.id))
noAnno.2 <- dic$jasparTOensembl[setdiff(jaspar$V2,dic$jasparTOtfclass$jaspar.id)]
knitr::kable(data.frame(jaspar.id=noAnno$V2,jaspar.name=noAnno$V3,
           ensembl.id= noAnno.2[noAnno$V2],
           url= paste0("http://jaspar2018.genereg.net/matrix/",noAnno$V2))%>%
  rowid_to_column('idx'))
```

## For TFclass, how many Jaspar Motifs found for each subfamily?
```{r,echo=F,warning=F,message=F}
stats.jaspar<-dic$merged%>%
  group_by(subfamily.id)%>%
  summarise(n.jaspar=length(unique(jaspar.id[!is.na(jaspar.id)])))%>%
  complete(subfamily.id,fill = list(n.jaspar=0))%>%
  arrange(as.character(subfamily.id))
stats.jaspar<- stats.jaspar[-1,]
cnt <- count(stats.jaspar[-nrow(stats.jaspar),],n.jaspar)%>%
  rename(n.tf.subfamilty=n)
knitr::kable(cnt)
```

## Which are those 178 subfamily have no jaspar motif?
```{r,echo=F,warning=F,message=F}
DT::datatable(dic$merged%>% 
                filter(subfamily.id %in% stats.jaspar$subfamily.id[stats.jaspar$n.jaspar==0])%>%
                dplyr::select(-starts_with("jaspar"))%>%
                dplyr::select(-starts_with("ensem"))%>%
                group_by(subfamily.id)%>%
                summarise_all(funs(paste(unique(.),collapse = ";")))%>%
                arrange(subfamily.id))
```

## Which are other subfamilies that found at least once in Jaspar?
```{r,echo=F,warning=F,message=F}
DT::datatable(
  stats.jaspar[stats.jaspar$n.jaspar>0,]%>% 
    right_join(dic$merged%>% 
                filter(subfamily.id %in% stats.jaspar$subfamily.id[stats.jaspar$n.jaspar>0])%>%
                dplyr::select(-starts_with("jaspar"))%>%
                dplyr::select(-starts_with("ensem"))%>%
                group_by(subfamily.id)%>%
                summarise_all(funs(paste(unique(.),collapse = ";"))))%>%
                arrange(subfamily.id)%>%
    drop_na())
```



